packages/core/index.ts

```diff
@@ -11,18 +11,35 @@ type Impossible<T extends Record<string, any>> = Partial<
   Record<keyof T, never>
 >;
 
-export interface BaseOptions<TShared extends Record<string, ZodType>> {
+export interface BaseOptions<
+  TShared extends Record<string, ZodType>,
+  TExtends extends Record<string, unknown> | undefined
+> {
   /**
    * How to determine whether the app is running on the server or the client.
    * @default typeof window === "undefined"
    */
   isServer?: boolean;
 
+  /**
+   * Another env that this env extends. This is useful for example when you have a "shared" env
+   * that is used by multiple apps and then app-specific envs for each app.
+   * This env must not have any conflicting keys with the extended env.
+   */
+  extends?: TExtends;
+
   /**
    * Shared variables, often those that are provided by build tools and is available to both client and server,
    * but isn't prefixed and doesn't require to be manually supplied. For example `NODE_ENV`, `VERCEL_URL` etc.
+   * Cannot overlap with extended env.
    */
-  shared?: TShared;
+  shared?: {
+    [TKey in keyof TShared]: TKey extends string
+      ? TKey extends keyof TExtends
```

Comment:
We can create a type util for this, to make it more shareable.

Reply:
Good idea. I've added a `DisallowDuplicateKeys` type helper. I was going to use it to DRY up the `ServerOptions`/`ClientOptions` types a bit, but doing so seems to cause some issues with type inference, so I'll keep those as they are for now.

Reply:
Yeah, understandable, that can come much later.